File: DoctorApp/src/main/java/SceneManager.java
import javafx.scene.Scene;
import javafx.scene.image.Image;
import javafx.stage.Stage;
import viewmodel.Config;

import java.sql.Connection;
import java.sql.Statement;

public class SceneManager {
    private static LogInPane logInPane = new LogInPane();
    private static DoctorMainPane doctorMainPane = new DoctorMainPane();
    private static DoctorRecordPane doctorRecordPane = new DoctorRecordPane();
    private static Stage stage = null;
    public static Connection conn = null;
    public static Statement statement = null;
    public static String selected = null;

    public static void init(){
        Scene[] scenes = {logInPane.scene, doctorMainPane.scene, doctorRecordPane.scene};
        logInPane.init();
        doctorRecordPane.init();
        doctorMainPane.init();
        addStylesheets(scenes);
    }

    public static void addStylesheets(Scene[] scenes){
        for (Scene scene: scenes) {
            scene.getStylesheets().add(Config.CSS_STYLES);
        }
    }

    public static void setStage(Stage stage_arg) {
        stage = stage_arg;
        stage.setTitle("MediRec");
        stage.getIcons().add(new Image("images/icon.png"));
        stage.setOnCloseRequest(event -> {
            System.out.println("exiting....");
        });
    }

    private static void showScene(Scene scene_arg){
        if (stage == null)
            return;
        stage.hide();
        stage.setScene(scene_arg);
        stage.show();
    }

    public static void showLogInScene(){
        showScene(logInPane.scene);
    }

    public static void showMainMenuScene(){
        showScene(doctorMainPane.scene);
    }

    public static void updateList() {
        doctorMainPane.updateList();
    }

    public static void showRecordScene(){
        doctorRecordPane.loadrecords(selected);
        selected = null;
        showScene(doctorRecordPane.scene);
    }



}

File: ValidatorClient/src/main/kotlin/viewmodel/SceneManager.kt
package viewmodel

import javafx.scene.Scene
import javafx.scene.image.Image
import javafx.stage.Stage
import main.Helper
import viewmodel.panes.*
import java.io.File
import kotlin.text.Charsets.UTF_8

object SceneManager {
    private val sizingIndex = if(Config.WIDTH < 1500) 1.5 else 2.0
    private val logInScene = Scene(LogInPane, Config.WIDTH / sizingIndex, Config.HEIGHT / sizingIndex)
    private val mainMenuScene = Scene(MainMenuPane, Config.WIDTH, Config.HEIGHT)
    private val scanScene = Scene(ScanPane, Config.WIDTH, Config.HEIGHT)
    private val qrScene = Scene(QRCodePane, Config.WIDTH, Config.HEIGHT)
    private val hospitalScene = Scene(ViewHospitalsPane, Config.WIDTH, Config.HEIGHT)
    private val addHospitalScene = Scene(AddHospitalPane, Config.WIDTH / sizingIndex, Config.HEIGHT / sizingIndex)
    private val addRemoveValidatorScene = Scene(AddRemoveValidatorsPane, Config.WIDTH / sizingIndex, Config.HEIGHT / sizingIndex)
    private val viewValidatorsScene = Scene(ViewValidatorsPane, Config.WIDTH, Config.HEIGHT)
    private val voteValidatorsScene = Scene(VoteValidatorsPane, Config.WIDTH, Config.HEIGHT)
    var stage: Stage? = null
        get() = field!!
        set(value) {
            if(value == null) return
            field = value
            field!!.title = "MediRec"
            field!!.icons.add(Image(Config.IMAGES["icon"]))
            field!!.setOnCloseRequest {
                println("Exiting...")
                (scanScene.root as ScanPane).disposeWebCamCamera()
                val file = File("${Config.BASE_PATH}/savedPatientsNotScanned.txt")
                if(Helper.nameToInfoMap.isNotEmpty()) {
                    val nameToInfoJsonString = StringBuilder("[")
                    Helper.nameToInfoMap.forEach { entry ->
                        nameToInfoJsonString.append("""{
                            "name": "${entry.key}",
                            "info": ${entry.value.first},
                            "pubKey": "${Helper.nameToPublicKey[entry.key]}",
                            "notFirst": ${entry.value.second}
                        },""")
                    }

                    file.writeText((nameToInfoJsonString.substring(0, nameToInfoJsonString.length - 1) + "]")
                            .replace(' ', '-')
                            .replace("\\s".toRegex(), ""), UTF_8)
                } else {
                    if(file.exists()) {
                        file.delete()
                    }
                }
            }
        }
    private var lastPatientInfoScene : Scene? = null
    var lastPatientPane : PatientInfoPane? = null

    init {
        val scenes = arrayOf(
                logInScene, mainMenuScene, scanScene,
                qrScene, hospitalScene, addHospitalScene,
                addRemoveValidatorScene, viewValidatorsScene, voteValidatorsScene
        )
        addStylesheets(*scenes)
    }

    private fun addStylesheets(vararg scenes: Scene) {
        for (scene in scenes) {
            scene.stylesheets.add(Config.CSS_STYLES)
        }
    }

    private fun showScene(scene: Scene) {
        if (stage == null)
            return
        stage!!.scene = scene
        if(!stage!!.isShowing)
            stage!!.show()
        maximize((scene != addHospitalScene && scene != logInScene && scene != addRemoveValidatorScene))
    }

    private fun maximize(maximize : Boolean) {
        stage!!.isMaximized = false
        if(maximize)
            stage!!.isMaximized = true
        else
            stage!!.sizeToScene()
    }

    fun showLogInScene() {
        showScene(logInScene)
    }

    fun showMainMenuScene() {
        MainMenuPane.selectLast()
        showScene(mainMenuScene)
    }

    fun showScanScene(type: ScanPane.TYPE) {
        showScene(scanScene)
        (scanScene.root as ScanPane).startWebCam()
        (scanScene.root as ScanPane).type = type
    }

    fun showQRScene() {
        QRCodePane.showButtons()
        showScene(qrScene)
    }

    fun showPatientInfoScene(infoPane: PatientInfoPane) {
        val scene = Scene(infoPane, Config.WIDTH, Config.HEIGHT)
        lastPatientPane = infoPane
        lastPatientInfoScene = scene
        addStylesheets(scene)
        showScene(scene)
    }

    fun showHospitalViewScene() {
        showScene(hospitalScene)
    }

    fun showAddHospitalScene() {
        showScene(addHospitalScene)
    }

    fun showAddRemoveValidatorScene() {
        showScene(addRemoveValidatorScene)
    }

    fun showVoteValidatorsScene() {
        VoteValidatorsPane.loadList()
        showScene(voteValidatorsScene)
    }

    fun showViewValidatorsScene() {
        ViewValidatorsPane.loadList()
        showScene(viewValidatorsScene)
    }

}
